[DOC] How to test and write benchmarks (#4016)
scikit-image/scikit-image: CONTRIBUTING.txt

----

Please `report bugs on GitHub <https://github.com/scikit-image/scikit-image/issues>`_.

Benchmarks
----------

While not mandatory for most pull requests, we ask that performance related
PRs include a benchmark in order to clearly depict the use-case that is being
optimized for. In this section we will review how to setup the benchmarks,
and three commands ``asv dev``, ``asv run`` and ``asv continuous``.

Prerequisites
~~~~~~~~~~~~~
Begin by installing `airspeed velocity <https://asv.readthedocs.io/en/stable/>`_
in your development environment. Prior to installation, be sure to activate your
development environment, then if using ``venv`` you may install the requirement with::

  source skimage-dev/bin/activate
  pip install asv

If you are using conda, then the command::

  conda activate skimage-dev
  conda install asv

is more appropriate. Once installed, it is useful to run the command::

  asv machine

To let airspeed velocity know more information about your machine.

Writing a benchmark
~~~~~~~~~~~~~~~~~~~
To write  benchmark, add a file in the ``benchmarks`` directory which contains a

> a class with a ``setup`` method and a ``time_`` method.

Reviewer: ```suggestion
a class with one ``setup`` method and at least one method prefixed with ``time_``.
```

The setup method should contain all the code necessary to prepare your data
before the function you wish to benchmark. The ``time_`` method

> should contain calls to the functions you wish to benchmark.

Reviewer: I like your structure but I think this part could be more explicit. Maybe something like
```
The ``time_`` method should only contain code you wish to benchmark. 
Therefore it is useful to move everything that prepares the benchmark scenario
into the ``setup`` method. This function is called before calling a ``time_``
method and its execution time is not factored into the benchmarks.
```

Author: Thanks!

For example, the ``setup`` method may include loading data into memory, and
potentially the allocation of memory for results. We take for example the
``TransformSuite`` benchmark:

.. code-block:: python

  import numpy as np
  from skimage import transform

  class TransformSuite:
      """Benchmark for transform routines in scikit-image."""

      def setup(self):
          self.image = np.zeros((2000, 2000))
          idx = np.arange(500, 1500)
          self.image[idx[::-1], idx] = 255
          self.image[idx, idx] = 255

      def time_hough_line(self):
          result1, result2, result3 = transform.hough_line(self.image)

> Here, the creation of the image is completed is in the ``setup`` method, and not

Reviewer 2: I do not understand this sentence, I guess there is a typo.

Author: I just want to bring to the attention of the contributor that things like creating our loading the dataset should not happen in the time function.

Reviewer: That's a good thought but may be redundant with my suggestion above. 
```suggestion
Here, the creation of the image is completed in the ``setup`` method, and not
```

Reviewer 2: I prefer @lagru suggestion then. It's easier to understand for me.

Reviewer 2: Then, it will be ready to be merged on my side.

included in the reported time of the benchmark.

It is also possible to benchmark features such as peak memory usage. To learn

> more about features of `asv`, we refer the read to the official

Reviewer: ```suggestion
more about the features of `asv`, please refer to the official
```

`airpseed velocity documentation <http://asv.readthedocs.io/en/latest/writing_benchmarks.html>`_.

Testing the benchmarks locally
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Prior to running the true benchmark, it is often worthwhile to test that the

> code if free of typos. To do so, you may use the ``asv dev``::

Reviewer: ```suggestion
code is free of typos. To do so, you may use the ``asv dev``::
```

  asv dev -b TransformSuite

Where the ``TransformSuite`` above will be run once in your current environment
to test that everything is in order.

Running your benchmark
~~~~~~~~~~~~~~~~~~~~~~

> The command above is fast, but doesn't always test the performance of the code

Reviewer: ```suggestion
The command above is fast, but doesn't test the performance of the code
```
I don't think the ``dev`` command is ever appropriate to benchmark.

adequately. To do that you may want to run the benchmark in your current
environment to see the performance of your change as you are developing new

> features. the command ``asv run -E existing`` will specify that you wish to run

Reviewer 2: ```suggestion
features. The command ``asv run -E existing`` will specify that you wish to run
```

the benchmark in your existing environment. This will save a significant amount
of time since building scikit-image can be a time consuming task::

  asv run -E existing -b TransformSuite

Comparing results to master
~~~~~~~~~~~~~~~~~~~~~~~~~~~

Often, the goal of a PR is to compare the results of the modifications in terms
speed to a snapshot of the code that is in the master branch of the

> ``scikit-image`` repository. The ``asv continuous`` is of help here::

Reviewer: ```suggestion
``scikit-image`` repository. The command ``asv continuous`` is of help here::
```
